File: ios/Packages/OpenUILang/Sources/OpenUILang/RuntimeValue.swift
import Foundation

/// A plain object with JS-like insertion-ordered keys (last-write keeps the
/// original key position, like JS object assignment).
struct RTObject {
    /// Own keys in INSERTION order — the raw slot list, before
    /// `OrdinaryOwnPropertyKeys` hoists the canonical array indices.
    private(set) var insertionKeys: [String] = []
    /// Keyed on `JSKey` (UTF-16 code-unit identity): JS object properties are
    /// code-unit exact, and program text can produce canonically-equal but
    /// code-unit-different keys (string-literal object keys, e.g. NFC vs NFD
    /// "café") that must stay distinct.
    private var storage: [JSKey: RTValue] = [:]

    /// The object's `[[Prototype]]`. A plain object literal starts at
    /// `Object.prototype`; `assign`ing `"__proto__"` re-points it, and `.null`
    /// means a genuinely prototype-less object.
    ///
    /// This slot is what makes lang-core's duck-typing (`isASTNode`,
    /// `isElementNode`, `containsDynamicValue`, the serializer's `isAstNode`)
    /// see through a `{"__proto__": …}` entry — see `JSObjects`.
    var prototype: RTValue = .proto(.object)

    init() {}
    init(_ pairs: [(String, RTValue)]) {
        for (k, v) in pairs { self[k] = v }
    }

    subscript(key: String) -> RTValue? {
        get { storage[JSKey(key)] }
        set {
            let k = JSKey(key)
            guard let newValue else {
                if storage.removeValue(forKey: k) != nil {
                    insertionKeys.removeAll { jsStringEquals($0, key) }
                }
                return
            }
            if storage[k] == nil { insertionKeys.append(key) }
            storage[k] = newValue
        }
    }

    /// Own keys in `Object.keys(o)` / `OrdinaryOwnPropertyKeys` order — every
    /// canonical array index FIRST in ascending numeric order, then the rest in
    /// insertion order.
    ///
    /// This is the runtime object, and its key order IS tree-observable: an
    /// `@Each` template that captures a row object runs it through
    /// `toLiteralAST` → `Obj.entries` → the `$ast` entries ARRAY, which the
    /// serializer emits verbatim (arrays are never sorted). Fixture
    /// `083-runtime-object-key-order`.
    var keys: [String] { jsOwnPropertyKeys(insertionKeys) }

    var entries: [(key: String, value: RTValue)] {
        keys.map { ($0, storage[JSKey($0)]!) }
    }
    var values: [RTValue] { keys.map { storage[JSKey($0)]! } }
    func has(_ key: String) -> Bool { storage[JSKey(key)] != nil }
    var isEmpty: Bool { insertionKeys.isEmpty }

    /// The one key JS object ASSIGNMENT never turns into an own property.
    static let protoKey = JSObjects.protoKey

    /// JS plain-object ASSIGNMENT (`o[key] = value`), which is NOT the same as
    /// defining an own property.
    ///
    /// A fresh `{}` inherits `Object.prototype`'s `__proto__` ACCESSOR, so
    /// `o["__proto__"] = v` invokes that setter, and the setter does NOT merely
    /// drop the key: for an object (or `null`) `v` it RE-POINTS the receiver's
    /// `[[Prototype]]`, which the whole duck-typing layer then reads through
    /// (fixtures `080-proto-object-key`, `085`–`087`). For a primitive `v` it
    /// does nothing at all. Either way `"__proto__"` never becomes an own
    /// property, so it never shows up in `Object.keys` / `JSON.stringify`.
    ///
    /// When the receiver no longer inherits that accessor (its chain was cut
    /// with `"__proto__": null`), assignment falls back to creating an ordinary
    /// own property.
    ///
    /// Use this at every site whose JS original is an assignment. Sites whose
    /// original is `Object.fromEntries` / `CreateDataPropertyOrThrow` keep
    /// using the subscript — those DO create an own `__proto__`
    /// (evaluator.js:40).
    mutating func assign(_ key: String, _ value: RTValue) {
        if key == RTObject.protoKey, JSObjects.inheritsProtoAccessor(.object(self)) {
            // ES 20.1.2.11 set Object.prototype.__proto__: object or null
            // re-points the prototype, anything else is a silent no-op.
            if value.isObjectOrFunction {
                prototype = value
            } else if case .null = value {
                prototype = .null
            }
            return
        }
        self[key] = value
    }
}

/// An element node in the materialized/evaluated tree
/// (lang-core `ElementNode`).
struct RTElement {
    var typeName: String
    var props: RTObject
    var partial: Bool
    var hasDynamicProps: Bool
    var statementId: String? = nil
}

/// A dynamically-typed runtime value mirroring the JS value space that flows
/// through materialization and evaluation (spec/openui-lang.md §8–§9).
/// `undefined` and `null` are distinct: `undefined` entries are omitted from
/// serialized objects while `null` entries are kept.
indirect enum RTValue {
    case undefined
    case null
    case bool(Bool)
    case number(Double)
    case string(String)
    case array([RTValue])
    case object(RTObject)
    case element(RTElement)
    /// A leftover AST node (builtin call, runtime expression, deferred slot).
    case ast(ASTNode)
    /// A native function value, reachable by inheriting one from a prototype
    /// (`$obj.toString`, `$obj.constructor`). `typeof` is `"function"`, not
    /// `"object"`, and `JSON.stringify` drops it exactly like `undefined`.
    case function(name: String, arity: Int)
    /// One of the intrinsic prototype objects, reachable through the
    /// `Object.prototype.__proto__` GETTER (`$obj.__proto__`).
    case proto(JSProtoKind)

    var isNullish: Bool {
        switch self {
        case .undefined, .null: return true
        default: return false
        }
    }

    /// `typeof v === "object" && v !== null` in JS terms (AST nodes, arrays,
    /// elements, intrinsic prototypes and plain objects are all objects there).
    ///
    /// `.function` is deliberately EXCLUDED: `typeof fn` is `"function"`, which
    /// is what gates `containsDynamicValue`, `evaluatePropCore`'s early return
    /// and its plain-object `needsEval` scan.
    var isObjectLike: Bool {
        switch self {
        case .array, .object, .element, .ast, .proto: return true
        default: return false
        }
    }

    /// `Type(v) is Object` — `isObjectLike` plus functions.
    var isObjectOrFunction: Bool {
        if case .function = self { return true }
        return isObjectLike
    }

    /// `JSON.stringify` omits an object property whose value is `undefined` OR
    /// a FUNCTION (ES 25.5.2 SerializeJSONProperty returns undefined for both).
    /// Inside an ARRAY both become `null` instead.
    var isDroppedByJSONStringify: Bool {
        switch self {
        case .undefined, .function: return true
        default: return false
        }
    }
}

/// Convert a plain JSON value (e.g. a schema `default`) into the runtime
/// value space. Object keys are inserted in sorted order — JSON object key
/// order is not observable downstream (the serializer sorts keys), sorting
/// just keeps the RTObject deterministic.
func jsonToRTValue(_ v: JSONValue) -> RTValue {
    switch v {
    case .null:
        return .null
    case .bool(let b):
        return .bool(b)
    case .number(let n):
        return .number(n)
    case .string(let s):
        return .string(s)
    case .array(let items):
        return .array(items.map { jsonToRTValue($0) })
    case .object(let o):
        var out = RTObject()
        for key in o.keys.sorted() {
            out[key] = jsonToRTValue(o[key]!)
        }
        return .object(out)
    }
}

// MARK: - JS coercion helpers

/// A JS `TypeError` escaping a prop evaluation.
///
/// `String(obj)` is `ToPrimitive(obj, string)`: call `obj.toString()` if
/// callable, else `obj.valueOf()` if callable, else throw
/// `TypeError: Cannot convert object to primitive value`. This value model has
/// no function values, so an own `toString` key (whatever it holds — a number,
/// a string, null) is never callable and the lookup falls through to
/// `Object.prototype.valueOf`, which returns the object itself and is rejected
/// as non-primitive. So a plain object with an own `"toString"` key ALWAYS
/// throws; without one, `Object.prototype.toString` answers `"[object Object]"`.
/// `"valueOf"` alone never matters at the string hint — `toString` is tried
/// first and succeeds.
///
/// `Evaluator.evaluateElementProps` catches this per prop, keeps the RAW prop
/// value and records a `runtimeErrors` entry, exactly like `evaluate-tree.js`'s
/// try/catch. Fixture `081-tostring-shadow-throws`.
struct JSTypeError: Error {
    /// The V8 message text, reproduced verbatim in the `runtimeErrors` entry.
    let message: String

    static let cannotConvertObjectToPrimitive =
        JSTypeError(message: "Cannot convert object to primitive value")
}

/// ECMAScript Number-to-String (shortest round-trip); reuses the serializer's
/// formatter for finite values.
func jsNumberToString(_ d: Double) -> String {
    if d.isNaN { return "NaN" }
    if d.isInfinite { return d > 0 ? "Infinity" : "-Infinity" }
    return TreeSerializer.formatNumber(d)
}

/// JS `Number(string)` semantics. Returns NaN for non-numeric strings.
/// ES *StrWhiteSpace* is exactly the `trim()` set (*WhiteSpace* ∪
/// *LineTerminator*; U+0085 NEL is NOT in it — former KNOWN-DEVIATION #7,
/// now fixed), so this reuses `jsTrim()` (Preprocess.swift), verified
/// empirically against node v22 `Number()` for every candidate scalar.
func jsStringToNumber(_ s: String) -> Double {
    let t = s.jsTrim()
    if t.isEmpty { return 0 }
    if t == "Infinity" || t == "+Infinity" { return .infinity }
    if t == "-Infinity" { return -.infinity }
    // Radix literals (no sign allowed).
    if t.hasPrefix("0x") || t.hasPrefix("0X") {
        return parseRadix(String(t.dropFirst(2)), radix: 16)
    }
    if t.hasPrefix("0o") || t.hasPrefix("0O") {
        return parseRadix(String(t.dropFirst(2)), radix: 8)
    }
    if t.hasPrefix("0b") || t.hasPrefix("0B") {
        return parseRadix(String(t.dropFirst(2)), radix: 2)
    }
    // Strict decimal literal: [+-]? (digits [. digits?] | . digits) ([eE][+-]?digits)?
    guard isStrictDecimalLiteral(t) else { return .nan }
    return Double(t) ?? .nan
}

private func parseRadix(_ digits: String, radix: Int) -> Double {
    if digits.isEmpty { return .nan }
    var value = 0.0
    for c in digits {
        guard let d = c.hexDigitValue, d < radix else { return .nan }
        value = value * Double(radix) + Double(d)
    }
    return value
}

private func isStrictDecimalLiteral(_ s: String) -> Bool {
    let chars = Array(s)
    var i = 0
    if i < chars.count && (chars[i] == "+" || chars[i] == "-") { i += 1 }
    var intDigits = 0
    while i < chars.count, chars[i].isASCII, chars[i].isNumber { intDigits += 1; i += 1 }
    var fracDigits = 0
    if i < chars.count && chars[i] == "." {
        i += 1
        while i < chars.count, chars[i].isASCII, chars[i].isNumber { fracDigits += 1; i += 1 }
    }
    if intDigits == 0 && fracDigits == 0 { return false }
    if i < chars.count && (chars[i] == "e" || chars[i] == "E") {
        i += 1
        if i < chars.count && (chars[i] == "+" || chars[i] == "-") { i += 1 }
        var expDigits = 0
        while i < chars.count, chars[i].isASCII, chars[i].isNumber { expDigits += 1; i += 1 }
        if expDigits == 0 { return false }
    }
    return i == chars.count
}

/// lang-core `toNumber`: number → itself (NaN passes through); numeric string
/// → number, non-numeric string → 0; boolean → 1/0; anything else → 0.
func dslToNumber(_ v: RTValue) -> Double {
    switch v {
    case .number(let n):
        return n
    case .string(let s):
        let n = jsStringToNumber(s)
        return n.isNaN ? 0 : n
    case .bool(let b):
        return b ? 1 : 0
    default:
        return 0
    }
}

/// JS `String(value)` semantics — including the case where it THROWS.
///
/// `String(obj)` is `ToPrimitive(obj, string)`: `toString` then `valueOf`, and
/// BOTH lookups go through the prototype chain (`JSObjects.getMember`), so the
/// answer depends on the receiver's `[[Prototype]]`, not just on its own keys:
///
/// - a plain object inherits `Object.prototype.toString` → `"[object Object]"`;
/// - an own `toString` key shadows it with a non-callable value (this model has
///   no user function values), so the lookup falls through to
///   `Object.prototype.valueOf`, which returns the object itself and is
///   rejected as non-primitive → THROWS (fixture `081-tostring-shadow-throws`);
/// - a prototype-LESS object (`{"__proto__": null, …}`) has neither method, so
///   it throws as well (fixture `087-proto-null-toprimitive-throws`) — the case
///   an own-key check alone can never see.
func jsToString(_ v: RTValue) throws -> String {
    switch v {
    case .undefined: return "undefined"
    case .null: return "null"
    case .bool(let b): return b ? "true" : "false"
    case .number(let n): return jsNumberToString(n)
    case .string(let s): return s
    case .array(let items):
        // Array.prototype.toString → join(","); null/undefined → "". A member
        // that shadows `toString` makes the join itself throw.
        return try items.map { $0.isNullish ? "" : try jsToString($0) }.joined(separator: ",")
    case .function(let name, _):
        return "function \(name)() { [native code] }"
    case .proto(let kind):
        // The intrinsic prototypes are ordinary objects of their own type:
        // Number.prototype is a Number object wrapping 0, String.prototype
        // wraps "", Boolean.prototype wraps false, Array.prototype is an empty
        // array and Function.prototype is an anonymous native function.
        switch kind {
        case .object: return "[object Object]"
        case .array: return ""
        case .number: return "0"
        case .string: return ""
        case .boolean: return "false"
        case .function: return "function () { [native code] }"
        }
    case .object, .element, .ast:
        return try objectToPrimitiveString(v)
    }
}

/// `ToPrimitive(v, string)` for the object-shaped runtime values: try
/// `toString`, then `valueOf`, each only if it resolves to a callable.
private func objectToPrimitiveString(_ v: RTValue) throws -> String {
    if let resolved = try JSObjects.getMemberWithOwner(v, "toString"),
        case .function = resolved.value
    {
        // `Array.prototype.toString` on a NON-array receiver reads `this.join`,
        // whose `this.length` is undefined → joins zero elements → "". Every
        // other reachable chain ends at `Object.prototype.toString`.
        if case .proto(.array) = resolved.owner { return "" }
        return "[object Object]"
    }
    // `valueOf` either is missing (prototype-less object) or is
    // `Object.prototype.valueOf`, which answers the object itself — never a
    // primitive. Both end the same way.
    throw JSTypeError.cannotConvertObjectToPrimitive
}

/// JS truthiness.
func jsTruthy(_ v: RTValue) -> Bool {
    switch v {
    case .undefined, .null: return false
    case .bool(let b): return b
    case .number(let n): return !(n == 0 || n.isNaN)
    case .string(let s): return !s.isEmpty
    case .array, .object, .element, .ast, .function, .proto: return true
    }
}

/// JS ToNumber for loose equality (differs from dslToNumber: non-numeric
/// strings yield NaN, null → 0, undefined → NaN).
private func jsToNumberStrict(_ v: RTValue) throws -> Double {
    switch v {
    case .undefined: return .nan
    case .null: return 0
    case .bool(let b): return b ? 1 : 0
    case .number(let n): return n
    case .string(let s): return jsStringToNumber(s)
    case .array, .object, .element, .ast, .function, .proto:
        return jsStringToNumber(try jsToString(v))
    }
}

/// JS loose equality (`==`).
func jsLooseEquals(_ a: RTValue, _ b: RTValue) throws -> Bool {
    switch (a, b) {
    case (.undefined, .undefined), (.undefined, .null), (.null, .undefined), (.null, .null):
        return true
    case (.undefined, _), (.null, _), (_, .undefined), (_, .null):
        return false
    case (.number(let x), .number(let y)):
        return x == y
    case (.string(let x), .string(let y)):
        // JS compares strings by UTF-16 code units — canonically equivalent
        // NFC/NFD variants are NOT equal (Swift's `==` would say they are).
        return jsStringEquals(x, y)
    case (.bool(let x), .bool(let y)):
        return x == y
    case (.bool(let x), _):
        return try jsLooseEquals(.number(x ? 1 : 0), b)
    case (_, .bool(let y)):
        return try jsLooseEquals(a, .number(y ? 1 : 0))
    case (.number(let x), .string(let y)):
        return x == jsStringToNumber(y)
    case (.string(let x), .number(let y)):
        return jsStringToNumber(x) == y
    default:
        // object-vs-primitive: ToPrimitive(object) → string, then compare.
        if a.isObjectLike && !b.isObjectLike {
            return try jsLooseEquals(.string(try jsToString(a)), b)
        }
        if !a.isObjectLike && b.isObjectLike {
            return try jsLooseEquals(a, .string(try jsToString(b)))
        }
        // KNOWN-DEVIATION (README.md #3): JS object == object is reference
        // identity; the port's value semantics have no identities, so this is
        // uniformly false — which matches the oracle, since materialized
        // values are always freshly built distinct objects.
        return false
    }
}

File: android/openui-lang/src/main/kotlin/dev/appless/openuilang/RuntimeValue.kt
package dev.appless.openuilang

/**
 * A plain JS-like object: insertion-ordered keys, last write keeps the
 * original key position (JS object assignment semantics).
 *
 * Unlike the Swift port, no `JSKey` wrapper is needed: JVM `String.equals` /
 * `hashCode` are UTF-16 code-unit based, so a `LinkedHashMap` already keeps
 * canonically-equivalent-but-distinct keys apart (precomposed vs decomposed
 * "café" — fixture 072).
 */
internal class RtObject() {
    private val map = LinkedHashMap<String, RtValue>()

    constructor(pairs: List<Pair<String, RtValue>>) : this() {
        for ((k, v) in pairs) put(k, v)
    }

    val keys: List<String> get() = map.keys.toList()
    val entries: List<Pair<String, RtValue>> get() = map.entries.map { it.key to it.value }
    val values: List<RtValue> get() = map.values.toList()
    val isEmpty: Boolean get() = map.isEmpty()

    operator fun get(key: String): RtValue? = map[key]

    fun put(key: String, value: RtValue) {
        map[key] = value
    }

    operator fun set(key: String, value: RtValue): Unit = put(key, value)

    fun has(key: String): Boolean = map.containsKey(key)

    companion object {
        fun of(vararg pairs: Pair<String, RtValue>): RtObject = RtObject(pairs.toList())
    }
}

/** An element node in the materialized/evaluated tree (lang-core `ElementNode`). */
internal class RtElement(
    val typeName: String,
    val props: RtObject,
    val partial: Boolean,
    val hasDynamicProps: Boolean,
    val statementId: String? = null,
) {
    fun withStatementId(id: String?): RtElement =
        RtElement(typeName, props, partial, hasDynamicProps, id)

    fun withProps(newProps: RtObject): RtElement =
        RtElement(typeName, newProps, partial, hasDynamicProps, statementId)
}

/**
 * A dynamically-typed runtime value mirroring the JS value space that flows
 * through materialization and evaluation (spec/openui-lang.md §8–§9).
 * `Undefined` and `Null` are distinct: `undefined` entries are omitted from
 * serialized objects while `null` entries are kept.
 */
internal sealed interface RtValue {
    data object Undefined : RtValue
    data object Null : RtValue
    data class Bool(val value: Boolean) : RtValue
    data class Num(val value: Double) : RtValue
    data class Str(val value: String) : RtValue
    data class Arr(val items: List<RtValue>) : RtValue
    data class Obj(val obj: RtObject) : RtValue
    data class Element(val element: RtElement) : RtValue

    /** A leftover AST node (builtin call, runtime expression, deferred slot). */
    data class Ast(val node: AstNode) : RtValue
}

internal val RtValue.isNullish: Boolean
    get() = this is RtValue.Undefined || this is RtValue.Null

/** `typeof v === "object" && v !== null` in JS terms. */
internal val RtValue.isObjectLike: Boolean
    get() = this is RtValue.Arr || this is RtValue.Obj ||
        this is RtValue.Element || this is RtValue.Ast

/**
 * Convert a plain JSON value (e.g. a schema `default`) into the runtime value
 * space. Object keys are inserted sorted — JSON key order is not observable
 * downstream (the serializer sorts), this just keeps the result deterministic.
 */
internal fun jsonToRtValue(v: JsonValue): RtValue = when (v) {
    is JsonValue.Null -> RtValue.Null
    is JsonValue.Bool -> RtValue.Bool(v.value)
    is JsonValue.Num -> RtValue.Num(v.value)
    is JsonValue.Str -> RtValue.Str(v.value)
    is JsonValue.Arr -> RtValue.Arr(v.value.map { jsonToRtValue(it) })
    is JsonValue.Obj -> {
        val out = RtObject()
        for (key in v.value.keys.sortedWith(JS_STRING_ORDER)) {
            out[key] = jsonToRtValue(v.value.getValue(key))
        }
        RtValue.Obj(out)
    }
}

// ── JS coercion helpers ─────────────────────────────────────────────────────

/** ECMAScript `Number::toString` (shortest round-trip), incl. non-finite. */
internal fun jsNumberToString(d: Double): String = when {
    d.isNaN() -> "NaN"
    d.isInfinite() -> if (d > 0) "Infinity" else "-Infinity"
    else -> TreeSerializer.formatNumber(d)
}

/**
 * JS `Number(string)` semantics — returns NaN for non-numeric strings.
 *
 * Hand-rolled rather than delegated to `String.toDoubleOrNull()`, which is
 * `Double.parseDouble` and accepts a pile of things JS rejects: trailing
 * `d`/`D`/`f`/`F` suffixes (`"1f"` → 1.0, JS → NaN), hex-float literals
 * (`"0x1p3"`), and `Character.isWhitespace` padding. It also REJECTS things
 * JS accepts (`"0x10"` → 16, `"0b101"` → 5, `"Infinity"`).
 *
 * ES *StrWhiteSpace* is exactly the [jsTrim] set.
 */
internal fun jsStringToNumber(s: String): Double {
    val t = s.jsTrim()
    if (t.isEmpty()) return 0.0
    if (t == "Infinity" || t == "+Infinity") return Double.POSITIVE_INFINITY
    if (t == "-Infinity") return Double.NEGATIVE_INFINITY
    // Radix literals (no sign allowed).
    if (t.length > 2 && t[0] == '0') {
        when (t[1]) {
            'x', 'X' -> return parseRadix(t.substring(2), 16)
            'o', 'O' -> return parseRadix(t.substring(2), 8)
            'b', 'B' -> return parseRadix(t.substring(2), 2)
        }
    }
    if (!isStrictDecimalLiteral(t)) return Double.NaN
    return t.toDoubleOrNull() ?: Double.NaN
}

private fun parseRadix(digits: String, radix: Int): Double {
    if (digits.isEmpty()) return Double.NaN
    var value = 0.0
    for (c in digits) {
        val d = when (c) {
            in '0'..'9' -> c - '0'
            in 'a'..'f' -> c - 'a' + 10
            in 'A'..'F' -> c - 'A' + 10
            else -> return Double.NaN
        }
        if (d >= radix) return Double.NaN
        value = value * radix + d
    }
    return value
}

private fun isStrictDecimalLiteral(s: String): Boolean {
    var i = 0
    if (i < s.length && (s[i] == '+' || s[i] == '-')) i++
    var intDigits = 0
    while (i < s.length && s[i] in '0'..'9') { intDigits++; i++ }
    var fracDigits = 0
    if (i < s.length && s[i] == '.') {
        i++
        while (i < s.length && s[i] in '0'..'9') { fracDigits++; i++ }
    }
    if (intDigits == 0 && fracDigits == 0) return false
    if (i < s.length && (s[i] == 'e' || s[i] == 'E')) {
        i++
        if (i < s.length && (s[i] == '+' || s[i] == '-')) i++
        var expDigits = 0
        while (i < s.length && s[i] in '0'..'9') { expDigits++; i++ }
        if (expDigits == 0) return false
    }
    return i == s.length
}

/**
 * lang-core `toNumber` (spec §9.1): number → itself (NaN passes through);
 * numeric string → number, non-numeric string → 0; boolean → 1/0; anything
 * else → 0.
 */
internal fun dslToNumber(v: RtValue): Double = when (v) {
    is RtValue.Num -> v.value
    is RtValue.Str -> {
        val n = jsStringToNumber(v.value)
        if (n.isNaN()) 0.0 else n
    }

    is RtValue.Bool -> if (v.value) 1.0 else 0.0
    else -> 0.0
}

/** JS `String(value)` semantics. */
internal fun jsToString(v: RtValue): String = when (v) {
    is RtValue.Undefined -> "undefined"
    is RtValue.Null -> "null"
    is RtValue.Bool -> if (v.value) "true" else "false"
    is RtValue.Num -> jsNumberToString(v.value)
    is RtValue.Str -> v.value
    // Array.prototype.toString → join(","); null/undefined → "".
    is RtValue.Arr -> v.items.joinToString(",") { if (it.isNullish) "" else jsToString(it) }
    is RtValue.Obj, is RtValue.Element, is RtValue.Ast -> "[object Object]"
}

/** JS truthiness. */
internal fun jsTruthy(v: RtValue): Boolean = when (v) {
    is RtValue.Undefined, is RtValue.Null -> false
    is RtValue.Bool -> v.value
    is RtValue.Num -> !(v.value == 0.0 || v.value.isNaN())
    is RtValue.Str -> v.value.isNotEmpty()
    else -> true
}

/**
 * JS loose equality (`==`), spec §9.1.
 *
 * KNOWN-DEVIATION (mirrors the Swift port's #3): object-vs-object comparison
 * is reference identity in JS; this port has value semantics and no stable
 * identities, so object == object is uniformly `false` — which is what the
 * oracle produces too, because the materializer always builds freshly
 * distinct objects.
 */
internal fun jsLooseEquals(a: RtValue, b: RtValue): Boolean {
    if (a.isNullish || b.isNullish) return a.isNullish && b.isNullish
    if (a is RtValue.Num && b is RtValue.Num) return a.value == b.value
    // JS compares strings by UTF-16 code units — String.equals on the JVM is
    // exactly that (canonically-equivalent NFC/NFD variants are NOT equal).
    if (a is RtValue.Str && b is RtValue.Str) return a.value == b.value
    if (a is RtValue.Bool && b is RtValue.Bool) return a.value == b.value
    if (a is RtValue.Bool) return jsLooseEquals(RtValue.Num(if (a.value) 1.0 else 0.0), b)
    if (b is RtValue.Bool) return jsLooseEquals(a, RtValue.Num(if (b.value) 1.0 else 0.0))
    if (a is RtValue.Num && b is RtValue.Str) return a.value == jsStringToNumber(b.value)
    if (a is RtValue.Str && b is RtValue.Num) return jsStringToNumber(a.value) == b.value
    // object-vs-primitive: ToPrimitive(object) → string, then compare.
    if (a.isObjectLike && !b.isObjectLike) return jsLooseEquals(RtValue.Str(jsToString(a)), b)
    if (!a.isObjectLike && b.isObjectLike) return jsLooseEquals(a, RtValue.Str(jsToString(b)))
    return false
}
